File: scripts/check-risk-paths.py
#!/usr/bin/env python3
"""Classify changed files into governance-sensitive buckets.

This script is a lightweight projection of:
- docs/GOVERNANCE.md
- docs/ARCHITECTURE.md
- docs/TEST_STRATEGY.md

It is intentionally warning-oriented by default. It does not replace
canonical policy documents.
"""

from __future__ import annotations

import argparse
import fnmatch
import os
import subprocess
import sys
from dataclasses import dataclass
from typing import Iterable


DOCS = {
    "governance": "docs/GOVERNANCE.md",
    "architecture": "docs/ARCHITECTURE.md",
    "test": "docs/TEST_STRATEGY.md",
    "grid": "docs/GRID_SCREEN_STANDARD.md",
    "permission": "docs/MENU_ACTION_PERMISSION_PLAN.md",
    "ledger": "docs/TASK_LEDGER.md",
}


@dataclass(frozen=True)
class Rule:
    key: str
    title: str
    patterns: tuple[str, ...]
    risk: str
    why: str
    how_to_fix: str
    docs: tuple[str, ...]


RULES: tuple[Rule, ...] = (
    Rule(
        key="grid_changed",
        title="Shared Grid / registry change",
        patterns=(
            "frontend/src/components/grid/**",
            "frontend/src/lib/grid/**",
            "config/grid-screens.json",
        ),
        risk="R2",
        why="Shared Grid contracts and screen registry changes can affect multiple screens.",
        how_to_fix=(
            "Run frontend grid validation and record regression evidence. "
            "Recommended: cd frontend && npm run validate:grid && npm run lint && npm run build"
        ),
        docs=(DOCS["grid"], DOCS["test"], DOCS["governance"]),
    ),
    Rule(
        key="permission_changed",
        title="Menu / permission path change",
        patterns=(
            "backend/app/api/menu.py",
            "backend/app/services/menu_service.py",
            "backend/app/schemas/menu.py",
        ),
        risk="R2",
        why="Permission-related changes must be checked on both UI and server sides.",
        how_to_fix=(
            "Confirm approval if scope is broader than local behavior, then verify both UI behavior "
            "and server-side enforcement."
        ),
        docs=(DOCS["permission"], DOCS["governance"], DOCS["test"]),
    ),
    Rule(
        key="auth_changed",
        title="Authentication / authorization change",
        patterns=(
            "backend/app/api/auth.py",
            "backend/app/core/auth.py",
            "backend/app/schemas/auth.py",
            "backend/app/services/auth_service.py",
        ),
        risk="R3",
        why="Authentication and authorization are high-risk protected areas.",
        how_to_fix=(
            "Require explicit approval, run auth/permission regression checks, and attach evidence "
            "before merging."
        ),
        docs=(DOCS["governance"], DOCS["test"], DOCS["ledger"]),
    ),
    Rule(
        key="payroll_changed",
        title="Payroll semantic path change",
        patterns=(
            "backend/app/api/payroll_phase2.py",
            "backend/app/services/payroll_phase2_service.py",
            "backend/app/schemas/payroll_phase2.py",
        ),
        risk="R3",
        why="Payroll-related changes can affect financial meaning and require stronger review.",
        how_to_fix=(
            "Require explicit approval, verify representative seed cases, and document rollback or stop conditions."
        ),
        docs=(DOCS["governance"], DOCS["test"], DOCS["ledger"]),
    ),
    Rule(
        key="deploy_changed",
        title="Deploy / infra path change",
        patterns=(
            ".github/workflows/deploy.yml",
            "docker-compose.deploy.yml",
            "backend/Dockerfile",
            "frontend/Dockerfile",
        ),
        risk="R3",
        why="Deploy and infra changes directly affect the current main-based operating flow.",
        how_to_fix=(
            "Require explicit approval, describe main impact, and prepare a health-check or rollback note."
        ),
        docs=(DOCS["governance"], DOCS["ledger"]),
    ),
)


def normalize(path: str) -> str:
    return path.replace("\\", "/")


def match_any(path: str, patterns: Iterable[str]) -> bool:
    path = normalize(path)
    return any(fnmatch.fnmatch(path, pattern) for pattern in patterns)


def git(*args: str) -> str:
    result = subprocess.run(["git", *args], capture_output=True, text=True)
    if result.returncode != 0:
        raise RuntimeError(result.stderr.strip() or "git command failed")
    return result.stdout.strip()


def resolve_base_head(base: str | None, head: str | None) -> tuple[str, str]:
    resolved_head = head or git("rev-parse", "HEAD")
    resolved_base = base
    if not resolved_base:
        try:
            resolved_base = git("merge-base", "HEAD", "origin/main")
        except Exception:
            resolved_base = git("rev-parse", f"{resolved_head}~1")
    return resolved_base, resolved_head


def changed_files(base: str, head: str) -> list[str]:
    output = git("diff", "--name-only", base, head)
    if not output:
        return []
    return [normalize(line) for line in output.splitlines() if line.strip()]


def write_output(path: str | None, key: str, value: str) -> None:
    if not path:
        return
    with open(path, "a", encoding="utf-8") as f:
        f.write(f"{key}={value}\n")


def gha_warning(message: str) -> None:
    if os.environ.get("GITHUB_ACTIONS") == "true":
        print(f"::warning::{message}")


def main() -> int:
    parser = argparse.ArgumentParser(description="Classify changed files by governance-sensitive paths")
    parser.add_argument("--base", help="Base git ref/sha", default=None)
    parser.add_argument("--head", help="Head git ref/sha", default=None)
    parser.add_argument("--github-output", help="Path to GitHub Actions output file", default=None)
    parser.add_argument(
        "--fail-on-r3",
        action="store_true",
        help="Exit non-zero when any R3 path is changed. Default is warning-only.",
    )
    args = parser.parse_args()

    try:
        base, head = resolve_base_head(args.base, args.head)
        files = changed_files(base, head)
    except Exception as exc:
        print(f"[error] failed to classify changed files: {exc}", file=sys.stderr)
        return 2

    backend_changed = any(match_any(path, ("backend/**",)) for path in files)

    hits: dict[str, list[str]] = {rule.key: [] for rule in RULES}
    for path in files:
        for rule in RULES:
            if match_any(path, rule.patterns):
                hits[rule.key].append(path)

    r3_changed = bool(hits["auth_changed"] or hits["payroll_changed"] or hits["deploy_changed"])

    print("# Risk Path Classification")
    print(f"- Base: {base}")
    print(f"- Head: {head}")
    print(f"- Changed file count: {len(files)}")
    print()

    if files:
        print("## Changed Files")
        for path in files:
            print(f"- {path}")
        print()
    else:
        print("## Changed Files")
        print("- none")
        print()

    print("## Rule Summary")
    for rule in RULES:
        matched = hits[rule.key]
        status = "true" if matched else "false"
        print(f"- {rule.key}: {status}")
        if matched:
            print(f"  - title: {rule.title}")
            print(f"  - risk: {rule.risk}")
            print(f"  - matched paths: {', '.join(matched)}")
            print(f"  - why: {rule.why}")
            print(f"  - how to fix: {rule.how_to_fix}")
            print(f"  - docs: {', '.join(rule.docs)}")
            gha_warning(
                f"{rule.title} detected ({rule.risk}). Why: {rule.why} How to fix: {rule.how_to_fix} Docs: {', '.join(rule.docs)}"
            )
    print(f"- backend_changed: {'true' if backend_changed else 'false'}")
    print(f"- r3_changed: {'true' if r3_changed else 'false'}")

    write_output(args.github_output, "grid_changed", str(bool(hits["grid_changed"])).lower())
    write_output(args.github_output, "permission_changed", str(bool(hits["permission_changed"])).lower())
    write_output(args.github_output, "auth_changed", str(bool(hits["auth_changed"])).lower())
    write_output(args.github_output, "payroll_changed", str(bool(hits["payroll_changed"])).lower())
    write_output(args.github_output, "deploy_changed", str(bool(hits["deploy_changed"])).lower())
    write_output(args.github_output, "backend_changed", str(bool(backend_changed)).lower())
    write_output(args.github_output, "r3_changed", str(bool(r3_changed)).lower())

    if args.fail_on_r3 and r3_changed:
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/check-risk-paths.js
#!/usr/bin/env node
"use strict";

const childProcess = require("child_process");
const fs = require("fs");

const DOCS = {
  governance: "docs/GOVERNANCE.md",
  architecture: "docs/ARCHITECTURE.md",
  test: "docs/TEST_STRATEGY.md",
  grid: "docs/GRID_SCREEN_STANDARD.md",
  permission: "docs/MENU_ACTION_PERMISSION_PLAN.md",
  ledger: "docs/TASK_LEDGER.md",
};

const RULES = [
  { key: "grid_changed", title: "Shared Grid / registry change", patterns: ["frontend/src/components/grid/**", "frontend/src/lib/grid/**", "config/grid-screens.json"], risk: "R2", why: "Shared Grid contracts and screen registry changes can affect multiple screens.", how_to_fix: "Run frontend grid validation and record regression evidence. Recommended: cd frontend && npm run validate:grid && npm run lint && npm run build", docs: [DOCS.grid, DOCS.test, DOCS.governance] },
  { key: "permission_changed", title: "Menu / permission path change", patterns: ["backend-spring/src/main/java/com/vibehr/menu/**", "backend-spring/src/main/java/com/vibehr/**/Authorization.java", "frontend/src/app/api/menus/**"], risk: "R2", why: "Permission-related changes must be checked on both UI and server sides.", how_to_fix: "Confirm approval if scope is broader than local behavior, then verify both UI behavior and server-side enforcement.", docs: [DOCS.permission, DOCS.governance, DOCS.test] },
  { key: "auth_changed", title: "Authentication / authorization change", patterns: ["backend-spring/src/main/java/com/vibehr/auth/**", "backend-spring/src/main/java/com/vibehr/platform/security/**", "backend-spring/src/main/java/com/vibehr/**/Authorization.java", "backend-spring/src/test/java/com/vibehr/**/*AuthorizationTest.java", "backend-spring/src/test/java/com/vibehr/platform/security/**"], risk: "R3", why: "Authentication and authorization are high-risk protected areas.", how_to_fix: "Require explicit approval, run auth/permission regression checks, and attach evidence before merging.", docs: [DOCS.governance, DOCS.test, DOCS.ledger] },
  { key: "payroll_changed", title: "Payroll semantic path change", patterns: ["backend-spring/src/main/java/com/vibehr/payroll/**", "backend-spring/src/main/java/com/vibehr/time/**", "backend-spring/src/main/resources/mybatis/**", "backend-spring/src/test/java/com/vibehr/payroll/**", "backend-spring/src/test/java/com/vibehr/time/**"], risk: "R3", why: "Payroll-related changes can affect financial meaning and require stronger review.", how_to_fix: "Require explicit approval, verify representative seed cases, and document rollback or stop conditions.", docs: [DOCS.governance, DOCS.test, DOCS.ledger] },
  { key: "flyway_changed", title: "Flyway schema or reference-data change", patterns: ["backend-spring/src/main/resources/db/migration/**", "docs/spring-migration/flyway-*.json", "docs/spring-migration/flyway-*.md", "docs/spring-migration/flyway-*.sql", "scripts/spring-migration/flyway-*.js"], risk: "R3", why: "Flyway migration and reference-data ownership changes alter production schema or seeded permissions.", how_to_fix: "Require explicit approval, run frozen evidence and Flyway verification, and record candidate/rollback gates.", docs: [DOCS.governance, DOCS.test, DOCS.ledger] },
  { key: "deploy_changed", title: "Deploy / infra path change", patterns: [".github/workflows/**", "docker-compose*.yml", "backend-spring/Dockerfile", "frontend/Dockerfile", "backend-spring/src/main/resources/db/migration/**", ".env.deploy", "scripts/spring-migration/**"], risk: "R3", why: "Deploy, Flyway, and infrastructure changes directly affect the Java operating flow.", how_to_fix: "Require explicit approval, describe main impact, and prepare a health-check or rollback note.", docs: [DOCS.governance, DOCS.ledger] },
];

function normalize(value) { return String(value).replace(/\\/g, "/"); }
function patternMatches(value, pattern) {
  const escaped = normalize(pattern).replace(/[.+^${}()|[\]\\]/g, "\\$&").replace(/\*/g, ".*").replace(/\?/g, ".");
  return new RegExp(`^${escaped}$`).test(normalize(value));
}
function matchesAny(value, patterns) { return patterns.some((pattern) => patternMatches(value, pattern)); }
function git(...args) {
  const result = childProcess.spawnSync("git", args, { encoding: "utf8" });
  if (result.status !== 0) throw new Error((result.stderr || "git command failed").trim());
  return result.stdout.trim();
}
function resolveBaseHead(base, head) {
  const resolvedHead = head || git("rev-parse", "HEAD");
  if (base) return [base, resolvedHead];
  try { return [git("merge-base", "HEAD", "origin/main"), resolvedHead]; }
  catch { return [git("rev-parse", `${resolvedHead}~1`), resolvedHead]; }
}
function changedFiles(base, head) {
  const output = git("diff", "--name-only", base, head);
  return output ? output.split(/\r?\n/).filter(Boolean).map(normalize) : [];
}
function writeOutput(file, key, value) { if (file) fs.appendFileSync(file, `${key}=${value}\n`, "utf8"); }
function parseArguments(argumentsList) {
  const options = { base: null, head: null, githubOutput: null, failOnR3: false };
  for (let index = 0; index < argumentsList.length; index += 1) {
    const argument = argumentsList[index];
    if (argument === "--fail-on-r3") options.failOnR3 = true;
    else if (["--base", "--head", "--github-output"].includes(argument)) {
      if (!argumentsList[index + 1]) throw new Error(`${argument} requires a value`);
      const field = argument === "--github-output" ? "githubOutput" : argument.slice(2);
      options[field] = argumentsList[++index];
    } else throw new Error(`Unknown argument: ${argument}`);
  }
  return options;
}
function classify(files) {
  const hits = Object.fromEntries(RULES.map((rule) => [rule.key, []]));
  for (const file of files) for (const rule of RULES) if (matchesAny(file, rule.patterns)) hits[rule.key].push(file);
  return { hits, backend_changed: files.some((file) => matchesAny(file, ["backend-spring/**"])), r3_changed: Boolean(hits.auth_changed.length || hits.payroll_changed.length || hits.flyway_changed.length || hits.deploy_changed.length) };
}
function run(options) {
  const [base, head] = resolveBaseHead(options.base, options.head);
  const files = changedFiles(base, head);
  const result = classify(files);
  console.log("# Risk Path Classification");
  console.log(`- Base: ${base}`); console.log(`- Head: ${head}`); console.log(`- Changed file count: ${files.length}`); console.log("\n## Changed Files");
  console.log(files.length ? files.map((file) => `- ${file}`).join("\n") : "- none");
  console.log("\n## Rule Summary");
  for (const rule of RULES) {
    const matched = result.hits[rule.key];
    console.log(`- ${rule.key}: ${Boolean(matched.length)}`);
    if (matched.length) {
      console.log(`  - title: ${rule.title}`); console.log(`  - risk: ${rule.risk}`); console.log(`  - matched paths: ${matched.join(", ")}`); console.log(`  - why: ${rule.why}`); console.log(`  - how to fix: ${rule.how_to_fix}`); console.log(`  - docs: ${rule.docs.join(", ")}`);
      if (process.env.GITHUB_ACTIONS === "true") console.log(`::warning::${rule.title} detected (${rule.risk}). Why: ${rule.why} How to fix: ${rule.how_to_fix} Docs: ${rule.docs.join(", ")}`);
    }
  }
  console.log(`- backend_changed: ${result.backend_changed}`); console.log(`- r3_changed: ${result.r3_changed}`);
  for (const [key, value] of Object.entries({ grid_changed: Boolean(result.hits.grid_changed.length), permission_changed: Boolean(result.hits.permission_changed.length), auth_changed: Boolean(result.hits.auth_changed.length), payroll_changed: Boolean(result.hits.payroll_changed.length), flyway_changed: Boolean(result.hits.flyway_changed.length), deploy_changed: Boolean(result.hits.deploy_changed.length), backend_changed: result.backend_changed, r3_changed: result.r3_changed })) writeOutput(options.githubOutput, key, String(value));
  return result.r3_changed && options.failOnR3 ? 1 : 0;
}
function main() { try { process.exitCode = run(parseArguments(process.argv.slice(2))); } catch (error) { console.error(`[error] failed to classify changed files: ${error.message}`); process.exitCode = 2; } }
if (require.main === module) main();
module.exports = { RULES, classify, matchesAny, parseArguments, patternMatches, resolveBaseHead, run };
